Remove undesirable atoms

mathuin opened the pull request:
The previous version of the occupancy select code was incomplete and generated incorrect results.

The code was refactored and now generates more accurate results.

mathuin committed:
Changed tests to reflect actual results.

After consultation with the PIs, specific residues were identified as tests for the occupancy selection code.

mathuin committed:
Minor tweaks to tests.

Found a syntax error, some ugly whitespace, and decimal/float confusion.

mathuin committed:
Refactored occupancy select code.

Previous version did not remove undesirable atoms.

BioPython has the Select class for this kind of task.

Rewrote the code to pass the tests.

mathuin committed:
Minor fixes, rebuilt fixtures.

Rebuilt structure after occupancy selection, removed unnecessary altloc selection, and added ability to delete residues by old_id.  All of this necessitated rebuilding the fixture.

mathuin committed:
Cleaned up tests.

This approach allows for additional proteins to be tested.

pop commented on pgd_splicer/ProcessPDBTask.py:
```diff
@@ -403,6 +403,58 @@ def atom_noamino(s):
+    """
+    This class implements the occupancy awareness selection.
+
+    Each disordered residue contains one or more disordered atom structures.  These structures contain one or more atoms, each with its own altloc and occupancy values.  As each atom associated with a particular altloc has its own occupancy value, the altloc with the highest average occupancy value is identified.  Atoms with other altloc values are removed from the structure.
```
pep8 violation, 80 characters per line.

mathuin replied:
This file has 173 PEP8 violations, most of which are more interesting than line-length.

I will wrap the comments. :-)

pop replied:
Heh, I had no idea. The actual source of my gripe was that GH was cutting off the line so I couldn't read the whole thing. Thanks for changing it though :-)

pop commented on pgd_splicer/ProcessPDBTask.py:
```diff
@@ -403,6 +403,58 @@ def atom_noamino(s):
+
+    def accept_residue(self, residue):
+        """
+        This method is being overloaded to perform the calculations required to properly accept atoms in the model based on the highest average occupancy value per altloc.
```
pep8 violation, 80 characters per line.

mathuin replied:
Fixed.

pop commented on pgd_splicer/ProcessPDBTask.py:
```diff
@@ -403,6 +403,58 @@ def atom_noamino(s):
+                            occ_list[altloc] = [occ]
+
+            best_altlocs = sorted(occ_list, key=lambda k: sum(occ_list[k])/len(occ_list[k]), reverse=True)
+            self.best_atoms[residue] = {atom: next(altloc for altloc in best_altlocs if altloc in res_occ[atom] or ' ') for atom in res_occ}
```
These lines, if you can, should also be limited to 80 characters. If it breaks readability then don't bother.

mathuin replied:
These can't go under 80 without getting super ugly or breaking things.  Renaming variables would probably be the only way to get smaller, and that falls under ugly IMO.

pop commented on pgd_splicer/ProcessPDBTask.py:
```diff
@@ -403,6 +403,58 @@ def atom_noamino(s):
+
+    def accept_atom(self, atom):
+        """
+        This method uses the information generated by accept_residue and stored in best_altlocs to accept only those atoms which are not disordered or with an altloc that matches the best altloc for that particular set of atoms.
```
_cough_ 80 characters per line _cough_

mathuin replied:
Fixed.

pop commented on pgd_splicer/ProcessPDBTask.py:
```diff
@@ -755,6 +789,9 @@ def parseWithBioPython(code, props, chains_filter=None):
                 prev       = None
                 if res_id in residues:
                     del residues[res_id]
+                if 'old_id' in vars():
```
Let's add a comment about what `vars()` is so future devs and future us knows it's a python thing and not a Django or PGD thing.

mathuin replied:
Added.

mathuin committed:
Minor modifications for line length and clarity.

Comments were too long, so I wrapped them.

Commands that were too long lose coherency so I left them.

Documented the vars() trick for future me and others.

mathuin committed:
Added new structure for tests.

1EJG was observed to not work with the new code, so here it is.

The test as committed fails, which is to be expected.  Now I have to fix it!

mathuin committed:
Fixed occupancy code again.

Because the occupancy requirements cross residue boundaries, the check
is done at the chain level.  Shims have been added to resolve invalid
residue issues at an earlier state, which may improve performance.

mathuin committed:
Reordered occupancy code, plus debugging.

Not all disordered atoms were being examined.  This has been corrected.

The res_occ variable was being overwritten, causing loss of information.

Some debugging that was helpful was left in but commented.  In the future, perhaps a logger should be passed to the class.

mathuin committed:
Added check for structure parsing.

mathuin committed:
Added logging support.

Debugging this is hard without logging.

mathuin committed:
Added argparse support for logging.

Relevant changes:
 * ProcessPDB.log is now optional.
 * Input only accepted via stdin at the moment.

mathuin committed:
Enabled logging and pre-cleaning.

The logging feature is now enabled from the command line.

The hetflags and not-in-AA3to1 detection and removal is now enabled.

Finally, the altlocs of all kept atoms are set to ' '.

mathuin committed:
Fixed missing atom check.

The code that checked for missing atoms was rewritten to catch cases where the best altloc choice was incomplete.  Less missing-atom residues will result.

mathuin committed:
Properly trapped no-acceptable-altloc case.

Sometimes no altloc is good.  Let's count those and report back.

mathuin committed:
Disable debugging code.

Forgot to disable the writing of postocc PDB files.

mathuin committed:
Fix for B factors, plus BioPython update.

BioPython made a subtle change to their DSSP code, so we have to change
as well.  Also, the B factor issue reported by Dale should be fixed.

mathuin committed:
Fixed division for B factors.

B factors are real, so they should use real division, not floor division.

See code.o.o #18261

mathuin committed:
Fixed occupancy like B factor.

The same issue impacting B factor with respect to hydrogens is also
affecting occupancy.  This will fix that issue.

GiriB committed:
Merge pull request #1 from mathuin/remove-undesirable-atoms

Merge mathuin/pgd:Remove-undesirable-atoms into GiriB/pgd:Remove-undesirable-atoms

GiriB committed:
Fix logging issue

Adds a FileHandler and StreamHandler for separate logging
to File and Console respectively.

mathuin committed:
Merge pull request #2 from GiriB/remove-undesirable-atoms

Separate logging to File and Console

mathuin committed:
Removed hydrogen atoms.

The PGDSelect class is used for selecting the most appropriate atoms
based on occupancy.  It has been modified to ignore hydrogen atoms when
performing this calculation and to not accept those atoms when
pre-processing the residues.  This should remove hydrogen atoms from
the database.

mathuin committed:
Moved PGDSelect into its own file.

The PGDSelect class was getting big and hard to maintain inside the
main script so I put it in its own file.  The AA3to1 check was disabled
as it no longer worked without easy access to that variable.

mathuin committed:
Handled case where H is only disordered atom.

Turns out sometimes the only atoms that are disordered are hydrogens
and since we're removing those, there's no reason to consider those
residues disordered.

mathuin committed:
continue not break doh

mathuin committed:
Cleaner handling of ordered residues.

Another side-effect of stripping hydrogen from residues and handling
those residues for which they are no longer disordered.

Also improved debugging for protein, chain and residue models.

mathuin committed:
Added support for icodes in select code.

Some residues have icodes -- 199A instead of just 199.

The old code did not differentiate between residues based on icodes.

Now it does!

mathuin committed:
Refactored AA_CHOICES to use BioPython values.

BioPython has translations between three-letter residue types
and single-letter codes.  So does PGD.  I removed PGD's version
and used BioPython's version instead.  Fewer lines of code, same
functionality.

mathuin committed:
Removed hydrogen checks.

Now that we're stripping hydrogen atoms in the select, we can
stop lokoing for them in parseWithBioPython.

mathuin committed:
Disabled old prefilter, added is_hydrogen method.

The hydrogen check was changed to look for hydrogen or deuterium.  It was cleaner
to add it as its own method at this point.

Previously we prefiltered ATOM lines that had no valid amino acids and HETATM lines
with valid amino acids.  This prefiltering is no longer necessary and has an negative
impact on processing efficiency.

mathuin committed:
Removed MAINTAINER line.
